Main.hs

```diff
@@ -1,29 +1,79 @@
+
 module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
 
 data Expr
+  = Num Double
+  | Sqrt Expr
+  | Add Expr Expr
+  | Sub Expr Expr
+  | Mul Expr Expr
+  | Div Expr Expr
+  | Pow Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Num x) = show x
+  show (Sqrt e) = "sqrt(" ++ show e ++ ")"
+  show (Add e1 e2) = "(" ++ show e1 ++ " + " ++ show e2 ++ ")"
+  show (Sub e1 e2) = "(" ++ show e1 ++ " - " ++ show e2 ++ ")"
+  show (Mul e1 e2) = "(" ++ show e1 ++ " * " ++ show e2 ++ ")"
+  show (Div e1 e2) = "(" ++ show e1 ++ " / " ++ show e2 ++ ")"
+  show (Pow e1 e2) = "(" ++ show e1 ++ " ^ " ++ show e2 ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (Num x) == (Num y) = x == y
+  (Sqrt e1) == (Sqrt e2) = e1 == e2
+  (Add e1 e2) == (Add f1 f2) = e1 == f1 && e2 == f2
+  (Sub e1 e2) == (Sub f1 f2) = e1 == f1 && e2 == f2
+  (Mul e1 e2) == (Mul f1 f2) = e1 == f1 && e2 == f2
+  (Div e1 e2) == (Div f1 f2) = e1 == f1 && e2 == f2
+  (Pow e1 e2) == (Pow f1 f2) = e1 == f1 && e2 == f2
+  _ == _ = False
 
 data Error
+  = DivByZero
+  | NegativeSqrt
```

Comment:
It would also be nice to be able to report which sub-expression caused the errors.